056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+ 3. Force rolls a broker if the pod is in stuck state
+ 4. Rolls a broker if the configuration has changed and cannot be updated dynamically
+ 5. Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+
+#### Unready pod
+If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck and considers to force roll. A pod is considered stuck if it is in one of following states:
+- `CrashLoopBackOff`
+- `ImagePullBackOff`
+- `ContainerCreating`
+- `Pending` and `Unschedulable`
+
+If pod is not stuck but unready, the next rollability checks are performed.
+
+
+#### Configuration changes
+ - Retrieves the current Kafka configurations of the broker via admin client and compares it with the desired configurations specified in the Kafka CR. 
+ - Performs dynamic configuration updates if possible
+
+In KRaft mode the KafkaRoller currently skips controller only nodes, but performs the above steps on any combined or broker only nodes.
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+1. Unready controller/combined nodes
+2. Ready controller/combined nodes
+3. Active controller (applies to both pure controller and combined node)
+4. Unready broker-only nodes
+5. Ready broker-only nodes
+
+
+### Triggers
+The following are some of the triggers that would roll a KRaft controller or combined node:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready 
+- Controller's configuration has changed
+
+The triggers for broker remain the same as Zookeeper mode.
+
+### Rollability checks
+The checks made by the KafkaRoller in different modes is described below.
+
+#### The new quorum check
+The checks include a new check for controllers to verify that rolling the node does not affect the quorum health.
+The proposed check is:
+- Create admin client connection to the brokers and call `describeMetadataQuorum` API
+- If failed to connect to the brokers, return `UnforceableProblem` which will result in delay and retry for the pod until the maximum attempts is reached.
```

Comment:
What is the timeout? Is there some kind of readiness check? When the cluster breaks, you do not want to wait for some long timeouts to fix things.

Reply:
We would be using the existing logic for handling `UnforceableProblem` and retrying the controller to see if it can connect to the brokers. It shouldn't wait for a long time. There is a readiness check at the beginning, for every pod.  If the controllers still can't connect to the brokers after the max num of retries, KafkaRoller would try reconcile the brokers same as usual.